File: Intatis-Windows/src/Intatis.Core/Session/EventLog.cs
using System.Text;
using System.Text.Json.Nodes;
using Intatis.Core.Protocol;

namespace Intatis.Core.Session;

public sealed class EventLogException : Exception
{
    public string Code { get; }

    public EventLogException(string code, string message) : base(message) => Code = code;
}

/// <summary>
/// Append-only JSONL session store. Append is the only mutation; replay and stream are
/// projections. Sequence numbers are monotonic and gap-free from zero per session, and
/// the writer lease guarantees a single runtime per session file across processes.
/// </summary>
public sealed class EventLog : IDisposable
{
    private readonly object _gate = new();
    private FileStream? _writerLock;
    private FileStream? _dataStream;
    private StreamWriter? _dataWriter;
    private long _lastSeq = -1;
    private bool _disposed;

    public string SessionId { get; }
    public string FilePath { get; }
    public string SessionDirectory { get; }

    /// <summary>Raised after an envelope has been durably appended.</summary>
    public event Action<Envelope>? EnvelopeAppended;

    private EventLog(string sessionId, string file, string directory)
    {
        SessionId = sessionId;
        FilePath = file;
        SessionDirectory = directory;
    }

    public static EventLog Open(string sessionId, string file)
    {
        var directory = Path.GetDirectoryName(file)!;
        Directory.CreateDirectory(directory);

        // Writer lease: exclusive handle on the sidecar lock; a second runtime fails closed.
        var lockPath = file + ".writer.lock";
        FileStream writerLock;
        try
        {
            // OpenOrCreate + FileShare.None: a live second runtime conflicts on the share
            // mode, while a stale lock file left by a crashed process opens cleanly.
            writerLock = new FileStream(lockPath, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.None);
        }
        catch (IOException)
        {
            throw new EventLogException("writer_already_active",
                $"another runtime already owns the writer lease for {sessionId}");
        }

        var log = new EventLog(sessionId, file, directory) { _writerLock = writerLock };
        log.RescanTail();
        return log;
    }

    public static bool HasActiveWriter(string file)
        => File.Exists(file + ".writer.lock")
           && ExceptionsOnExclusiveOpen(file + ".writer.lock");

    private static bool ExceptionsOnExclusiveOpen(string path)
    {
        try
        {
            using var probe = File.Open(path, FileMode.Open, FileAccess.ReadWrite, FileShare.None);
            return false;
        }
        catch (IOException)
        {
            return true;
        }
    }

    private void RescanTail()
    {
        _lastSeq = -1;
        if (!File.Exists(FilePath)) return;
        using var stream = new FileStream(FilePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
        using var reader = new StreamReader(stream, Encoding.UTF8);
        string? line;
        while ((line = reader.ReadLine()) is not null)
        {
            if (line.Length == 0) continue;
            Envelope envelope;
            try { envelope = Envelope.FromJsonLine(line); }
            catch (Exception) { continue; } // fail-soft: skip undecodable lines (future types)
            if (envelope.Session != SessionId)
                throw new EventLogException("session_mismatch",
                    $"event line belongs to session {envelope.Session}, expected {SessionId}");
            if (envelope.Seq <= _lastSeq)
                throw new EventLogException("non_monotonic_sequence",
                    $"sequence regression at seq {envelope.Seq} (last {_lastSeq})");
            _lastSeq = envelope.Seq;
        }
    }

    public long LastSeq => _lastSeq;

    public Envelope Append(string type, JsonNode? payload, DateTime? ts = null, bool flush = true)
    {
        lock (_gate)
        {
            ObjectDisposedException.ThrowIf(_disposed, this);
            if (_dataWriter is null)
            {
                _dataStream = new FileStream(FilePath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Read);
                _dataStream.Seek(0, SeekOrigin.End);
                _dataWriter = new StreamWriter(_dataStream, Encoding.UTF8) { AutoFlush = false };
            }
            var envelope = new Envelope
            {
                Seq = _lastSeq + 1,
                Ts = ts ?? DateTime.UtcNow,
                Session = SessionId,
                V = 1,
                Type = type,
                Payload = payload?.DeepClone(),
            };
            _dataWriter.WriteLine(envelope.ToJsonLine());
            if (flush)
            {
                _dataWriter.Flush();
                _dataStream!.Flush(true);
            }
            _lastSeq = envelope.Seq;
            EnvelopeAppended?.Invoke(envelope);
            return envelope;
        }
    }

    public List<Envelope> Replay(long fromSeq = 0)
    {
        var result = new List<Envelope>();
        if (!File.Exists(FilePath)) return result;
        using var stream = new FileStream(FilePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
        using var reader = new StreamReader(stream, Encoding.UTF8);
        string? line;
        while ((line = reader.ReadLine()) is not null)
        {
            if (line.Length == 0) continue;
            Envelope envelope;
            try { envelope = Envelope.FromJsonLine(line); }
            catch (Exception) { continue; }
            if (envelope.Seq >= fromSeq) result.Add(envelope);
        }
        return result;
    }

    /// <summary>Replay, then follow live appends until the channel completes.</summary>
    public async IAsyncEnumerable<Envelope> StreamAllAsync(
        long fromSeq,
        [System.Runtime.CompilerServices.EnumeratorCancellation] CancellationToken ct = default)
    {
        foreach (var envelope in Replay(fromSeq))
            yield return envelope;

        var pending = System.Threading.Channels.Channel.CreateUnbounded<Envelope>();
        void Handler(Envelope e) => pending.Writer.TryWrite(e);
        EnvelopeAppended += Handler;
        try
        {
            while (!ct.IsCancellationRequested)
            {
                var next = await pending.Reader.ReadAsync(ct).ConfigureAwait(false);
                if (next.Seq >= fromSeq) yield return next;
            }
        }
        catch (OperationCanceledException) { /* normal stop */ }
        finally { EnvelopeAppended -= Handler; }
    }

    public void Dispose()
    {
        lock (_gate)
        {
            if (_disposed) return;
            _disposed = true;
            try { _dataWriter?.Flush(); } catch (IOException) { }
            _dataWriter?.Dispose();
            _dataWriter = null;
            _dataStream?.Dispose();
            _dataStream = null;
            _writerLock?.Dispose();
            _writerLock = null;
            try { File.Delete(FilePath + ".writer.lock"); } catch (IOException) { }
        }
    }
}

File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/session/EventLog.kt
package com.intatis.shared.session

import com.intatis.shared.protocol.Envelope
import kotlinx.serialization.json.JsonObject
import java.io.Closeable
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.RandomAccessFile
import java.nio.channels.FileLock
import java.time.Instant

class EventLogException(val code: String, message: String) : Exception(message)

/**
 * Append-only JSONL session store. Append is the only mutation; replay and stream are
 * projections. Sequence numbers are monotonic and gap-free from zero per session, and
 * the writer lease guarantees a single runtime per session file across processes.
 */
class EventLog private constructor(
    val sessionId: String,
    val filePath: File,
) : Closeable {

    private val lock = Any()
    private var writerLockHandle: RandomAccessFile? = null
    private var writerLock: FileLock? = null
    private var outputStream: FileOutputStream? = null
    private var lastSeq: Long = -1
    private var closed = false

    val sessionDirectory: File get() = filePath.parentFile!!

    /** Invoked on the writing thread after an envelope has been durably appended. */
    var onEnvelopeAppended: ((Envelope) -> Unit)? = null

    fun lastSequence(): Long = synchronized(lock) { lastSeq }

    companion object {
        fun open(sessionId: String, file: File): EventLog {
            file.parentFile?.mkdirs()

            // Writer lease: exclusive OS-level lock on the sidecar; a second runtime
            // fails closed. A stale lock file left by a crashed process locks cleanly.
            val lockPath = File(file.path + ".writer.lock")
            val handle = RandomAccessFile(lockPath, "rw")
            val fileLock = try {
                handle.channel.tryLock()
            } catch (e: java.nio.channels.OverlappingFileLockException) {
                handle.close()
                throw EventLogException("writer_already_active",
                    "another runtime owns ${lockPath.path}: overlapping lock in this process")
            } catch (e: IOException) {
                handle.close()
                throw EventLogException("writer_already_active", "another runtime owns ${lockPath.path}: ${e.message}")
            }
            if (fileLock == null) {
                handle.close()
                throw EventLogException("writer_already_active", "another runtime already owns the writer lease for $sessionId")
            }

            val log = EventLog(sessionId, file)
            log.writerLockHandle = handle
            log.writerLock = fileLock
            log.rescanTail()
            return log
        }

        fun hasActiveWriter(file: File): Boolean {
            val lockPath = File(file.path + ".writer.lock")
            if (!lockPath.exists()) return false
            return try {
                RandomAccessFile(lockPath, "rw").use { handle ->
                    handle.channel.tryLock() ?: return true
                }
                false
            } catch (_: java.nio.channels.OverlappingFileLockException) {
                true
            } catch (_: IOException) {
                true
            }
        }
    }

    private fun rescanTail() {
        lastSeq = -1
        if (!filePath.exists()) return
        filePath.forEachLine { line ->
            if (line.isEmpty()) return@forEachLine
            val envelope = try {
                Envelope.fromJsonLine(line)
            } catch (_: Exception) {
                return@forEachLine // fail-soft: skip undecodable lines (future types)
            }
            if (envelope.session != sessionId) {
                throw EventLogException("session_mismatch",
                    "event line belongs to session ${envelope.session}, expected $sessionId")
            }
            if (envelope.seq <= lastSeq) {
                throw EventLogException("non_monotonic_sequence",
                    "sequence regression at seq ${envelope.seq} (last $lastSeq)")
            }
            lastSeq = envelope.seq
        }
    }

    fun append(type: String, payload: JsonObject?, ts: Instant = Instant.now(), flush: Boolean = true): Envelope {
        synchronized(lock) {
            check(!closed) { "event log is closed" }
            val stream = outputStream ?: FileOutputStream(filePath, true).also { outputStream = it }
            val envelope = Envelope(
                seq = lastSeq + 1,
                ts = ts,
                session = sessionId,
                type = type,
                payload = payload,
            )
            stream.write((envelope.toJsonLine() + "\n").toByteArray(Charsets.UTF_8))
            if (flush) {
                stream.flush()
                stream.fd.sync()
            }
            lastSeq = envelope.seq
            onEnvelopeAppended?.invoke(envelope)
            return envelope
        }
    }

    fun replay(fromSeq: Long = 0): List<Envelope> {
        val result = mutableListOf<Envelope>()
        if (!filePath.exists()) return result
        synchronized(lock) {
            filePath.forEachLine { line ->
                if (line.isEmpty()) return@forEachLine
                val envelope = try {
                    Envelope.fromJsonLine(line)
                } catch (_: Exception) {
                    return@forEachLine
                }
                if (envelope.seq >= fromSeq) result.add(envelope)
            }
        }
        return result
    }

    override fun close() {
        synchronized(lock) {
            if (closed) return
            closed = true
            runCatching { outputStream?.flush() }
            runCatching { outputStream?.close() }
            outputStream = null
            runCatching { writerLock?.release() }
            runCatching { writerLockHandle?.close() }
            writerLock = null
            writerLockHandle = null
            runCatching { File(filePath.path + ".writer.lock").delete() }
        }
    }
}
